design/worker-pods.md

```diff
@@ -0,0 +1,161 @@
+# Worker Pods for Backup, Restore
+
+Status: Draft
+
+This document proposes a new approach to executing backups and restores where each operation is run in its own "worker pod" rather than in the main Velero server pod. This approach has significant benefits for concurrency, scalability, and observability.
+
+## Goals
+
+- Enable multiple backups/restores to be run concurrently by running each operation in its own worker pod.
+- Improve Velero's scalability by distributing work across multiple pods.
+- Allow logs for in-progress backups/restores to be streamed by the user.
+
+## Non Goals
+
+- Creating CronJobs for scheduled backups.
+
+## Background
+
+Today, the Velero server runs as a Kubernetes deployment with a single replica. The deployment's one pod runs all of the Velero controllers, including the backup and restore controllers. The backup and restore controllers are responsible for doing all of the work associated with backups/restores, including interacting with the Kubernetes API server, writing to or reading from object storage, creating or restoring disk snapshots, etc. 
+
+Because each Velero controller is configured to run with a single worker, only one backup or one restore can be executed at a time. This can lead to performance issues if a large number of backups/restores are triggered within a short time period. This issue will only be exacerbated as Velero moves towards a multi-tenancy model and better supports on-premises snapshot + backup scenarios.
+
+## High-Level Design
+
+Velero controllers will no longer directly execute backup/restore logic themselves (note: the rest of this document will refer only to backups, for brevity, but it applies equally to restores). Instead, when the controller is informed of a new backup custom resource, it will immediately create a new worker pod which is responsible for end-to-end processing of the backup, including validating the spec, scraping the Kubernetes API server for resources, triggering persistent volume snapshots, writing data to object storage, and updating the custom resource's status as appropriate.
+
+A worker pod will be given a deterministic name based on the name of the backup it's executing. This will prevent Velero from inadvertently creating multiple worker pods for the same backup, since any subsequent attempts to create a pod with the specified name will fail.
+
+This design trivially enables running multiple backups concurrently, as each one runs in its own isolated pod and the Velero server's backup controller does not need to wait for the backup to complete before spawning a worker pod for the next one. Additionally, Velero becomes much more scalable, as the resource requirements for the Velero server itself are largely unaffacted by the number of backups. Instead, Velero's scalability becomes limited only by the total amount of resources available in the cluster to process worker pods.
```

Comment:
Is remaining resource available something we would need to check for before initiating a backup? What happens if there's not enough resources to complete a backup? (I suppose the same questions would apply with todays design but the scenario is much less likely to happen).

Reply:
@carlisia I am still new to the project, but if the cluster did not have enough resources for the backup with this new design, I guess it will be reported in worker pod status? Assuming the worker pod would be unschedulable, instead of preemptively checking for cluster resources, reporting lack it from Pod API might be an easier alternative.